src/port_forwarding.rs

```diff
@@ -99,19 +106,22 @@ impl From<Box<Path>> for Socket<'static> {
 
 impl Socket<'_> {
     /// Create a new TcpSocket
-    pub fn new<T: ToSocketAddrs>(addr: &T) -> Result<Socket<'static>, io::Error> {
-        addr.to_socket_addrs()?
-            .next()
-            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "no more socket addresses to try"))
-            .map(Socket::TcpSocket)
+    pub fn new<'a, S>(host: S, port: u16) -> Socket<'a>
+    where
+        S: Into<Cow<'a, str>>,
```

Comment:
This changes the public api, and so would end up requiring a major version bump. Would you mind making this a separate constructor instead?

cc @NobodyXu we have to be a little careful about these kinds of changes. Maybe we should add a CI step that invokes https://github.com/obi1kenobi/cargo-semver-checks or https://github.com/Enselic/cargo-public-api to catch these things automatically (at least in most cases).

Reply:
> This changes the public api, and so would end up requiring a major version bump. Would you mind making this a separate constructor instead?

Sure, but with #131 we probably need a breaking change anyway.

> cc @NobodyXu we have to be a little careful about these kinds of changes. Maybe we should add a CI step that invokes https://github.com/obi1kenobi/cargo-semver-checks or https://github.com/Enselic/cargo-public-api to catch these things automatically (at least in most cases).

Yeah, we definitely need semver-checking in CI.

Reply:
I'm sorry to chime in, I saw the stream and think there's common mistake regarding SemVer and the zero major version you might not be aware of (and I learned recently as well). :-)

My comment is just for info, choose whatever version you want. I just want to give some hint.

> This changes the public api, and so would end up requiring a major version bump.

Technically, there's no need to bump the major version, because 1.0.0 hasn't been released yet. Therefore, everything so far is "initial development".

The following is cited from [SemVer spec](https://semver.org/#spec-item-4):

> Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

Reply:
Thanks!

Reply:
@stephan-ci in Rust, when the leftmost version number component is a 0, then the next digit over is effectively considered the major version. That's what I was referring to bumping. Updating x in 0.x has the same effect and implications as bumping x in x.0, and so it is still considered (and referred to) as a major version bump.

Reply:
Just a user here, so discount my opinion appropriately, but strong preference from me for a semver-breaking release with _just_ the new API, rather than a backcompatible release that includes both APIs! The existing API is, AFAICT, not possible to use correctly with remote port forwarding, and so forcing users into the new API seems worth the breakage to me.

(Earlier discussion on this: https://github.com/openssh-rust/openssh/pull/120#discussion_r1215492541)